flagscale/runner/backend.py

```diff
@@ -0,0 +1,2253 @@
+import copy
+import importlib
+import os
+import shlex
+import sys
+
+from abc import ABC, abstractmethod
+from datetime import datetime
+
+import hydra
+
+from hydra.core.hydra_config import HydraConfig
+from omegaconf import DictConfig, OmegaConf
+
+from flagscale.runner.runner_train import (
+    _get_args_megatron,
+    _get_args_pi0,
+    _get_args_robotics,
+    _update_config_train,
+)
+
+# from flagscale.runner.runner_base import RunnerBase
+from flagscale.runner.utils import (
+    flatten_dict_to_args,
+    flatten_dict_to_args_verl,
+    get_free_port,
+    get_nnodes,
+    get_nproc_per_node,
+    logger,
+    parse_hostfile,
+    run_local_command,
+    run_scp_command,
+    run_ssh_command,
+)
+from flagscale.serve.args_mapping.mapping import ARGS_CONVERTER
+
+
+def _get_args_ray(config: DictConfig):
+    # see the following link for more details
+    # https://github.com/facebookresearch/hydra/discussions/2750
+    config_dict = OmegaConf.to_container(config, resolve=True)
+
+    # step2: restructuring the config
```

Comment:
suggestion: instead of adding comments indicating what you are doing, please consider commenting the code by telling other WHY this `pop` operations are needed.